libpromises/evalfunction.c

```diff
@@ -7922,43 +8044,93 @@ static FnCallResult FnCallClassFilterData(
     }
 
-    /* Currently only parent type array is supported */
-    if (JsonGetType(parent) != JSON_TYPE_ARRAY)
-    {
-        JsonDestroyMaybe(parent, allocated);
-        return FnFailure();
-    }
-
     if (!allocated)
     {
+        /* Create heap allocated copy so that we can mutate it */
         parent = JsonCopy(parent);
         assert(parent != NULL);
     }
 
+    /* This argument can hold one of:
+     * - 'array_of_arrays'
+     * - 'array_of_objects'
+     * - 'object_of_objects'
+     * - 'auto' (to automatically detect the data structure)
+     */
     const char *data_structure = RlistScalarValue(args->next);
-    const char *key_or_index = RlistScalarValue(args->next->next);
 
-    /* Iterate through array backwards so we can avoid having to compute index
-     * offsets for each removed element */
-    for (size_t index_plus_one = JsonLength(parent); index_plus_one > 0; index_plus_one--)
-    {
-        assert(index_plus_one > 0);
-        size_t index = index_plus_one - 1;
-        JsonElement *child = JsonArrayGet(parent, index);
-        assert(child != NULL);
+    /* This argument can hold:
+     * - the index of the class expression in a child array
+     * - the key to the class expression in a child object
+     * - or NULL (the class expression is the key to the child object)
+     */
+    const char *key_or_index = (args->next->next != NULL)
+                             ? RlistScalarValue(args->next->next)
+                             : NULL;
+
+    JsonIterator iter;
+    JsonElement *child;
+    switch (JsonGetType(parent)) {
+        case JSON_TYPE_ARRAY:
+            /* Iterate through array in backwards order so that we can avoid
+             * having to compute index offsets for each removed element */
+            for (size_t index_plus_one = JsonLength(parent); index_plus_one > 0; index_plus_one--)
+            {
+                assert(index_plus_one > 0);
+                size_t index = index_plus_one - 1;
+                child = JsonArrayGet(parent, index);
+                assert(child != NULL);
 
-        bool remove;
-        if (!ClassFilterDataArray(ctx, fp->name, data_structure, key_or_index, child, &remove))
-        {
-            /* Error is already logged */
+                bool remove;
+                if (!ClassFilterDataArray(ctx, fp->name, data_structure, key_or_index, child, &remove))
+                {
+                    /* Error is already logged */
+                    JsonDestroy(parent);
+                    return FnFailure();
+                }
+
+                if (remove)
+                {
+                    JsonArrayRemoveRange(parent, index, index);
+                }
+            }
+            break;
+        case JSON_TYPE_OBJECT:
+            iter = JsonIteratorInit(parent);
+            child = JsonIteratorNextValue(&iter);
+
+            while ((child != NULL))
+            {
+                const char *exogenous_key = JsonIteratorCurrentKey(&iter);
```

Comment:
again, I think you should change exogenous_key to self_key or something that is easier to understand.